Adopt getParsedMembers()/getSemanticMembers() to squash nondeterminism

DougGregor opened the pull request:
Adopt `getParsedMembers()` and `getSemanticMembers()` liberally throughout the code base, eliminating most uses of the non-deterministic and ever-shifting `getMembers()`. Both of these queries provide deterministic results, so their adoption will eliminate a known sources of nondeterminism that affects Swift interface file generation, Swift module generation, the generated header, etc.

Addresses rdar://problem/59513440 and likely a number of other nondeterminism issues.

DougGregor committed:
Adopt getParsedMembers()/getSemanticMembers() in a number of places.

Switch most of the callers of `IterableDeclContext::getMembers()` over
to either `getParsedMembers()` or `getSemanticMembers()`, as
appropriate.

This eliminates several sources of nondeterminism in various
post-typechecking operations, such as interface generation, SIL
generation, and Swift module file generation.

DougGregor committed:
Include inferred associated type witnesses in the semantic members.

DougGregor committed:
Generalize the "semantic members" sort to all implicit members.

The "semantic members" query does a sort of synthesized functions to
ensure that they are provided in a stable order across translation
units. Generalize this sort operation to all implicit members, so that
`getSemanticMembers()` is deterministic.

While here, optimize the sort a bit so we only compute interface types
and turn them into strings in cases where the declaration type and
name are equivalent.

DougGregor committed:
Use "implicit" rather than "synthesized" bit for designated init check.

DougGregor committed:
[AST] Remove "synthesized" bit from AbstractFunctionDecl.

We're no longer using it for anything.

DougGregor committed:
[Semantic members] Loosen an assertion and cope with lazy properties.

DougGregor committed:
[Semantic members] Sort members in a semi-reasonable manner.

Rather than sorting implicit members by kind/name, provide a little bit
more structure, because this affects user-visible outputs like Swift
interfaces and generated headers.

The breadth of test-case updates here indicates the clients that have
changed: IR generation, interface printing, generic headers, module
generation, and SIL generation all become more deterministic and
intentional.

DougGregor commented:
@swift-ci please smoke test compiler performance

davidungar commented on include/swift/AST/Decl.h:
```diff
@@ -405,10 +405,6 @@ class alignas(1 << DeclAlignInBits) Decl {
     /// Whether the function body throws.
     Throws : 1,
 
-    /// Whether this member was synthesized as part of a derived
-    /// protocol conformance.
-    Synthesized : 1,
-
```
Hooray! One less bit!

davidungar commented on lib/AST/DeclContext.cpp:
```diff
@@ -804,10 +804,12 @@ ArrayRef<Decl *> IterableDeclContext::getParsedMembers() const {
       ctx.evaluator,
       SemanticMembersRequest{const_cast<IterableDeclContext *>(this)},
       ArrayRef<Decl *>());
+  assert(result.size() <= getMemberCount());
```
This assertion might deserve a comment: When would there ever be fewer semantic members than members? Are there noop members?

davidungar commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -2459,30 +2459,156 @@ NamingPatternRequest::evaluate(Evaluator &evaluator, VarDecl *VD) const {
+      case DeclKind::TypeAlias:
+        return SortedDeclKind::Type;
+
+      case DeclKind::Constructor:
+        return SortedDeclKind::Initializer;
+
+      case DeclKind::Destructor:
+        return SortedDeclKind::Deinitializer;
+
+      case DeclKind::Param:
+      case DeclKind::PatternBinding:
+      case DeclKind::Var:
+        return SortedDeclKind::Property;
+
+      case DeclKind::Accessor:
+        return SortedDeclKind::Function;
+
+      case DeclKind::Func:
+        return cast<FuncDecl>(valueDecl)->isOperator()
+            ? SortedDeclKind::Operator
+            : SortedDeclKind::Function;
+
+
+      case DeclKind::InfixOperator:
+      case DeclKind::PrecedenceGroup:
+      case DeclKind::PrefixOperator:
+      case DeclKind::PostfixOperator:
+      case DeclKind::Subscript:
+        return SortedDeclKind::Operator;
+
+      case DeclKind::Extension:
+      case DeclKind::IfConfig:
+      case DeclKind::Import:
+      case DeclKind::MissingMember:
+      case DeclKind::Module:
+      case DeclKind::TopLevelCode:
+      case DeclKind::PoundDiagnostic:
+        return SortedDeclKind::Other;
+    }
+  }
```
It looks like the only time one `DeclKind` -> two `SortedDeclKinds` is for `Func`. Would it be possible to just split the `DeclKind::Func` into `DeclKind:NamedFunc` and `DeclKind::Operator`? We could keep `DeclFunc` as an abstract class. Then, maybe we wouldn't need `SortedDeclKind` at all.

DougGregor replied:
That refactor isn't interesting to any other clients, and would not eliminate `SortedDeclKind` because we still want the ordering to be defined even if `DeclKind` gets refactored.

davidungar reviewed (approved):
Looks great! But I hope that rather than introducing `SortedDeclKind` it would be better to just split `DeclKind::Func` into `NamedFunc` and `OperatorFunc`. That way, you might be able to just sort on `DeclKind`.

DougGregor committed:
[Semantic members] Work around limitations of associated type inference.

When determining semantic members, do a bit of extra work to prune down
the set of associated types that might introduce typealiases through
associated type inference. Then, before triggering inference, compute
the interface types of all of the members in order---this works around
some limitations of associated type inference.

DougGregor committed:
[Code completion] Update the set of parsed members for delayed declarations.

Code completion delays the parsing of the declaration in which code
completion is happening, and then inserts it later. This happens
*after* we have populated the parse-members request, breaking the
model that this result is functional and immutable.

Patch up the result of the parse-members request when this happens, so
the rest of the frontend sees the correct (completed) result. This is a
localized hack to a more general problem: we shouldn't ever be
populating the parse-members request without including the delayed
declaration in it. However, addressing that is a more significant
refactoring.

DougGregor committed:
[Semantic members] Improve comment about forcing conformances

DougGregor commented:
@swift-ci smoke test

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci test source compatibility

DougGregor commented:
@swift-ci smoke test compiler performance

DougGregor commented:
@swift-ci please test source compatibility

DougGregor commented:
@swift-ci please smoke test compiler performance

DougGregor commented:
Jenkins fail on macOS. Trying again

DougGregor commented:
@swift-ci please smoke test macOS

shahmishal commented:
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))
